### DISCUSSION TITLE
Change the unit of the IFC file from mm to cm

### DISCUSSION THREAD

**Discussion by u0416033** [IfcOpenShell/IfcOpenShell#2737] [Q&A]
I have a IFC file with unit mm and I have to load it on the website with three.js . The IFC is too big so that out of the memory .I try to change the unit of the file and it is the code I try to use . Console outputs all of the attributes that unit will be affect by unit but the IFC file do not change with the code . How can I fix it?

//-------------------------------code [----------------------------------------]from(https://github.com/IfcOpenShell/IfcOpenShell/issues/1247)
```
ifc_file  = ifcopenshell.open('專案座標_東10000北5.ifc')


def calculate_unit_scale(file):
    units = file.by_type("IfcUnitAssignment")[0]
    unit_scale = 1
    for unit in units.Units:
        if not hasattr(unit, "UnitType") or unit.UnitType != "LENGTHUNIT":
            continue
        while unit.is_a("IfcConversionBasedUnit"):
            unit_scale *= unit.ConversionFactor.ValueComponent.wrappedValue
            unit = unit.ConversionFactor.UnitComponent
        if unit.is_a("IfcSIUnit"):
            unit_scale *= get_prefix_multiplier(unit.Prefix)
    return unit_scale


s = ifcopenshell.ifcopenshell_wrapper.schema_by_name("IFC4")
classes_to_modify = {}
for d in s.declarations():
    if not hasattr(d, "all_attributes") or "IfcLength" not in str(d.all_attributes()):
        continue
    attributes_to_modify = []
    for attribute in d.all_attributes():
        if "IfcLength" in str(attribute):
            attributes_to_modify.append(attribute.name())
    classes_to_modify[d.name()] = attributes_to_modify

print(classes_to_modify)



def scale_all(obj, sf):
    def serialize(obj):
        """Recursively walk object's hierarchy."""
        if isinstance(obj, (int, float)):
            return obj * sf
        elif isinstance(obj, list):
            return [serialize(item) for item in obj]
        elif isinstance(obj, tuple):
            return tuple(serialize([item for item in obj]))
        else:
            try:
                if obj.is_a('IfcLengthMeasure') is True:
                    obj.wrappedValue = obj.wrappedValue * sf
                    return obj
                elif obj.is_a('IfcReal') is True:
                    obj.wrappedValue = obj.wrappedValue * sf
                    return obj
                elif obj.is_a('IfcInteger') is True:
                    obj.wrappedValue = int(obj.wrappedValue * sf)
                    return obj
                elif obj.is_a('IfcPlaneAngleMeasure') is True:
                    obj.wrappedValue = obj.wrappedValue * sf
                    return obj
                elif obj.is_a('IfcText') is True:
                    return obj
                elif obj.is_a('IfcLogical') is True:
                    return obj
            except:
                pass

            raise ValueError(f'Unknown entity "{type(obj)}", "{obj}"')

    return serialize(obj)


for ifc_class, attributes in classes_to_modify.items():
    for element in ifc_file.by_type(ifc_class):
        for attribute in attributes:

            old_val = getattr(element, attribute)
            if old_val is None:
                continue
            setattr(element, attribute, scale_all(old_val, 10))
            new_val = getattr(element, attribute)
```
![image](https://user-images.githubusercontent.com/53708536/216776732-8ddc6c62-792f-494d-b143-10ac3686cecd.png)

**Comment by Moult** [comment]
You can use IfcPatch for this: https://blenderbim.org/docs-python/ifcpatch.html

One of the recipes converts length units: https://blenderbim.org/docs-python/autoapi/ifcpatch/recipes/ConvertLengthUnit/index.html

The underlying code is here: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcpatch/ifcpatch/recipes/ConvertLengthUnit.py

**Reply by u0416033** [reply]
Thanks! But Should I add IfcPatch to the "miniconda3\Lib\site-packages\ifcopenshell " and create the src folder for it?

**Reply by Moult** [reply]
Instructions are here: https://blenderbim.org/docs-python/ifcpatch.html#source-installation

**Reply by u0416033** [reply]
I don't know whether I need to clone all the source code or just the ifcpatch folder from : https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.7.0/src/ifcpatch
I just create the src folder and copy the ifcpatch folder in it
![image](https://user-images.githubusercontent.com/53708536/216816361-07a8436e-8aff-4daf-ac70-47c920bb3a4b.png)
![image](https://user-images.githubusercontent.com/53708536/216816579-ee83aff1-319e-4518-9f5a-ffd5cfe6978d.png)

**Reply by u0416033** [reply]
Sorry... I am not familiar with python

**Reply by Moult** [reply]
Just grab the IfcPatch folder :) You can either put it in your site-packages or in your current working directory, either will let Python find it.

**Reply by u0416033** [reply]
but it still have some error of it ,how can I fix it

![image](https://user-images.githubusercontent.com/53708536/216816786-a420ef29-6bce-4a71-b519-81520f6162a1.png)

I think python get the module but maybe something comes wrong?
![image](https://user-images.githubusercontent.com/53708536/216816841-c48ac9a9-915a-464e-8d06-68179de28a0c.png)

**Reply by aothms** [reply]
You have ifcpatch import as ifcopenshell.src.ifcpatch.ifcpatch, so when ifcpatch itself tries to import ifcpatch.recipes it fails.

instead you can try (untested):

```python
import os, sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), 'ifcopenshell', 'src', 'ifcpatch')))
import ifcpatch
```

**Reply by u0416033** [reply]
I have add the path I think it had work sucessfully but it has new problem below... Maybe it means that the ifcopenshell version is different to ifcpatch?

![image](https://user-images.githubusercontent.com/53708536/216831836-a10a7f5b-bfef-48a6-8eeb-41996ed460d8.png)

**Reply by aothms** [reply]
IfcPatch was recently changed, you now supply both input and file as arguments in the dictionary.

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcpatch/ifcpatch/__init__.py#L40-L43

**Reply by Moult** [reply]
Apologies I haven't regenerated the docs in a while, I've just done so.

**Reply by u0416033** [reply]
I see ,thanks very much! I want to make sure that what python version is this ifcpatch version (https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcpatch/ifcpatch/__init__.py#L40-L43) needed

**Reply by Moult** [reply]
Any IfcOpenShell supported Python version will work.

**Reply by u0416033** [reply]
Thank you very much .I have sucessfully import the ifcpatch but when I use the "recipe": "ConvertLengthUnit" it return "Attribute not set"
``
output=ifcpatch.execute({ 
    "input": "GRAPHISOFT_Archicad_Sample_Project_Hillside_House_v1.0_AC25.ifc",
    "file": ifcopenshell.open("GRAPHISOFT_Archicad_Sample_Project_Hillside_House_v1.0_AC25.ifc"),
    "recipe": "ConvertLengthUnit",
    "arguments": ["MILLIMETERS"]
  })
ifcpatch.write(output, "output.ifc")
``
Did I do something wrong of it ?
![image](https://user-images.githubusercontent.com/53708536/216916082-075cd48d-af32-4277-8f32-fa1797bbbf21.png)

![image](https://user-images.githubusercontent.com/53708536/216915843-312ffee5-f86c-423b-b89b-ebf7fe4519e5.png)

**Reply by aothms** [reply]
It's likely due to this, where the owner history is constructed.

https://github.com/IfcOpenShell/IfcOpenShell/blame/d028b52d77517d75eec7bacb4450cd32f7aca293/src/ifcopenshell-python/ifcopenshell/api/owner/create_owner_history.py#L103-L120

See here https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/owner/settings.py#L19 you are to overwrite get_user() and get_application()

**Reply by u0416033** [reply]
I see ,but how can I fix it ? Should I change the code of the ifcopenshell that you mentioned above?

**Reply by Moult** [reply]
Sorry, it was a bug that I didn't notice outside the Blender environment. https://github.com/IfcOpenShell/IfcOpenShell/commit/5879d1480346d2be4b18b5f6d58c5d9929aee86d

**Reply by u0416033** [reply]
Ok I see . Maybe I need to find other way to deal with this problem .Thank you a lot for replaying my questions!

**Reply by Moult** [reply]
@u0416033 did you see the commit? The bug has been fixed now https://github.com/IfcOpenShell/IfcOpenShell/commit/5879d1480346d2be4b18b5f6d58c5d9929aee86d

**Reply by u0416033** [reply]
![image](https://user-images.githubusercontent.com/53708536/217121695-969b1b79-cbc7-4db9-942c-38cfc02f0931.png)
![image](https://user-images.githubusercontent.com/53708536/217121713-54d51934-04af-4598-893e-13a6e27b9489.png)
Yes I have change the code of ConvertLengthUnit.py of ifcpatch but still have error in console.
![image](https://user-images.githubusercontent.com/53708536/217121925-1e047dcb-e68b-4232-afab-5fea5e24f544.png)
Is this error is ok ? Because it also return that " FINISHED  assign_unit " .  And the ifc file did not change .

**Reply by Moult** [reply]
Sorry, I seem to have overlooked that the model may contain monetary units. Fixed now: https://github.com/IfcOpenShell/IfcOpenShell/commit/3f8314bcf3176a52da431872d99e353925c57c6d

**Reply by u0416033** [reply]
Cool! I have successfully change the unit of it . I will try other ifcpatch function ! Thanks a lot!

**Reply by Moult** [reply]
Thanks for persisting and helping find bugs!

**Reply by u0416033** [reply]
@Moult Hi ,sorry .I still have some problem for the ifcpatch. I find that after I use the recipe of ConvertLengthUnit , the model does not modify correctly. Like the screenshot below.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/53708536/2dff9585-6563-4827-9f44-da2c0e2c4c79)
but it look like this originally
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/53708536/d4d52f8b-ae9d-409e-a3cf-24de81c436bf)

Can this problem be fixed?
my code :
`
ifc_file = ifcopenshell.open('南屯區永富段95、96地號測試檔(coRFR)-0731-mm.ifc')

output = ifcpatch.execute({
    "input": ifc_file,
    "file": ifc_file,
    "recipe": "ConvertLengthUnit",
    "arguments": ["CENTIMETERS"]
})
ifcpatch.write(output, "南屯區永富段95、96地號測試檔(coRFR)-0731-mm-modify.ifc")
`

**Reply by Moult** [reply]
Can you share your model? Send to dion@thinkmoult.com if private.

**Reply by u0416033** [reply]
@Moult  Hi ! Sorry for late reply,I have send the ifc file with email.

**Reply by u0416033** [reply]
@Moult 
Hello Moult. I am wondering if you have received the model file that I sent you. If so, were you able to run the function that corrects the model units? Thank you very much!

**Reply by Moult** [reply]
Sorry for being so late for this. I got your email and ran it. The arguments should be "CENTIMETER" with no S at the end. It seems to work with no errors on my end.
